common/src/main/java/org/apache/celeborn/common/network/server/ratelimit/RateLimitController.java

```diff
@@ -0,0 +1,251 @@
+    }
+
+    public long getTimestamp() {
+      return timestamp;
+    }
+
+    public BufferStatusHub getBufferStatusHub() {
+      return bufferStatusHub;
+    }
+  }
+
+  /**
+   * 1. If the total pending bytes is over high watermark, will congest users who produce speed is
+   * higher than the potential average consume speed.
+   *
+   * <p>2. Will stop congest these uses until the pending bytes lower to low watermark.
+   *
+   * <p>3. If the pending bytes doesn't exceed the high watermark, will allow all users to try to
+   * get max throughout capacity.
+   */
+  public boolean isUserCongested(UserIdentifier userIdentifier) {
+    if (userBufferStatuses.size() == 0) {
+      return false;
+    }
+
+    long pendingConsumed = getTotalPendingBytes();
+
+    // The potential consume speed in average
+    long avgConsumeSpeed = getPotentialConsumeSpeed();
+
+    if (!overHighWatermark.get() && pendingConsumed > highWatermark) {
+      logger.info(
+          "Pending consume bytes: {} higher than high watermark, need to congest it",
+          pendingConsumed);
+      overHighWatermark.set(true);
+    }
+
+    if (overHighWatermark.get()) {
+      if (pendingConsumed < lowWatermark) {
+        logger.info("Lower than low watermark, exit congestion control");
+        overHighWatermark.set(false);
+        return false;
+      }
+
+      UserBufferInfo userBufferInfo = userBufferStatuses.get(userIdentifier);
+      if (userBufferInfo != null) {
+        BufferStatusHub.BufferStatusNode userBufferStatus =
+            userBufferInfo.getBufferStatusHub().sum();
+
+        // If the user produce speed is higher that the avg consume speed, will congest it
+        long userProduceSpeed = userBufferStatus.numBytes() / sampleTimeWindowSeconds;
+        if (logger.isDebugEnabled()) {
+          logger.debug(
+              "The user {}, produceSpeed is {},"
+                  + " while consumeSpeed is {}, need to congest it: {}",
+              userIdentifier,
+              userProduceSpeed,
+              avgConsumeSpeed,
+              userProduceSpeed > avgConsumeSpeed);
+        }
+        return userProduceSpeed > avgConsumeSpeed;
+      }
+    }
+
+    return false;
+  }
+
+  public void incrementBytes(UserIdentifier userIdentifier, int numBytes) {
+    long currentTimeMillis = System.currentTimeMillis();
+    UserBufferInfo userBufferInfo =
+        userBufferStatuses.computeIfAbsent(
+            userIdentifier,
+            user -> {
+              logger.info("New user {} comes, initializing its rate status", user);
+              BufferStatusHub bufferStatusHub = new BufferStatusHub(sampleTimeWindowSeconds);
+              return new UserBufferInfo(currentTimeMillis, bufferStatusHub);
+            });
+
+    BufferStatusHub.BufferStatusNode node = new BufferStatusHub.BufferStatusNode(numBytes);
+    userBufferInfo.updateInfo(currentTimeMillis, node);
+  }
+
+  public void decrementBytes(int numBytes) {
```

Comment:
better to rename to consumeBytes, or it will be wired that inside a decrement method we actually add into BufferStatusNode